src/azure-cli-core/azure/cli/core/telemetry.py

```diff
@@ -4,269 +4,325 @@
+    _session.parameters = parameters
 
-def _user_agrees_to_telemetry():
-    return _get_azure_cli_config().getboolean('core', 'collect_telemetry', fallback=True)
 
+# definitions
 
-@_suppress_all_exceptions(fallback_return={})
+@decorators.call_once
+@decorators.suppress_all_exceptions(fallback_return={})
 def _get_azure_cli_config():
     from azure.cli.core._config import az_config
     return az_config
 
 
-@_suppress_all_exceptions(fallback_return=None)
+# internal utility functions
+
+@decorators.suppress_all_exceptions(fallback_return=None)
 def _get_core_version():
     from azure.cli.core import __version__ as core_version
     return core_version
 
 
-@_suppress_all_exceptions(raise_in_debug=True)
-def _safe_exec(props, key, fn):
-    props[key] = fn()
-
+@decorators.suppress_all_exceptions(fallback_return=None)
+def _get_installation_id():
+    return _get_profile().get_installation_id()
 
-@_suppress_one_exception(expected_exception=AttributeError, fallback_return=None)
-def _get_installation_id(profile):
-    try:
-        return profile.get_installation_id()
-    except AttributeError:
-        return None
 
-
-@_suppress_all_exceptions(fallback_return=None)
+@decorators.call_once
+@decorators.suppress_all_exceptions(fallback_return=None)
 def _get_profile():
     from azure.cli.core._profile import Profile
     return Profile()
 
 
-def _get_env_string():
-    return _remove_cmd_chars(_remove_symbols(str([v for v in os.environ
-                                                  if v.startswith('AZURE_CLI')])))
+@decorators.suppress_all_exceptions(fallback_return='')
+@decorators.hash256_result
+def _get_hash_mac_address():
+    s = ''
+    for index, c in enumerate(hex(uuid.getnode())[2:].upper()):
+        s += c
+        if index % 2:
+            s += '-'
 
+    s = s.strip('-')
 
-def _get_user_machine_id():
-    return _get_hash(platform.node() + getpass.getuser())
+    return s
 
 
-def _get_user_azure_id(profile):
-    try:
-        from azure.cli.core._util import CLIError
-    except ImportError:
-        CLIError = Exception
+@decorators.suppress_all_exceptions(fallback_return='')
+def _get_hash_machine_id():
+    # Definition: Take first 128bit of the SHA256 hashed MAC address and convert them into a GUID
+    return str(uuid.UUID(_get_hash_mac_address()[0:32]))
 
-    try:
-        return _get_hash(profile.get_current_account_user())
-    except (AttributeError, CLIError):  # pylint: disable=broad-except
-        return None
 
+@decorators.suppress_all_exceptions(fallback_return='')
+@decorators.hash256_result
+def _get_user_azure_id():
+    return _get_profile().get_current_account_user()
 
-def _get_hash(s):
-    hash_object = hashlib.sha256(s.encode('utf-8'))
-    return str(hash_object.hexdigest())
 
+def _get_env_string():
+    return _remove_cmd_chars(_remove_symbols(str([v for v in os.environ
+                                                  if v.startswith('AZURE_CLI')])))
```

Comment:
This is not a problem. We register merely the environment variable names not their values.